mono/mini/mini-amd64.c

```diff
@@ -2600,15 +2602,19 @@ mono_arch_start_dyn_call (MonoDynCallInfo *info, gpointer **args, guint8 *ret, g
 {
 	ArchDynCallInfo *dinfo = (ArchDynCallInfo*)info;
 	DynCallArgs *p = (DynCallArgs*)buf;
-	int arg_index, greg, freg, i, pindex;
+	int arg_index, greg, i, pindex;
 	MonoMethodSignature *sig = dinfo->sig;
 	int buffer_offset = 0;
-	static int param_reg_to_index [16];
+	static int general_param_reg_to_index[MONO_MAX_IREGS];
+	static int float_param_reg_to_index[MONO_MAX_FREGS];
+
 	static gboolean param_reg_to_index_inited;
 
 	if (!param_reg_to_index_inited) {
 		for (i = 0; i < PARAM_REGS; ++i)
-			param_reg_to_index [param_regs [i]] = i;
+			general_param_reg_to_index [param_regs[i]] = i;
+		for (i = 0; i < FLOAT_PARAM_REGS; ++i)
+			float_param_reg_to_index [float_param_regs[i]] = i;
 		mono_memory_barrier ();
```

Comment:
I recently learned, that this is not actually thread-safe.
The rule of thumb, is that one barrier is seldom/never enough. They come in pairs -- on the reader and the writer.
At least for ARM. So it'll only matter here, I guess, for an ARM[64]-hosted AMD64-targeted compiler which will be rare. I was surprised and still have trouble getting my head around it.

Can this data therefore be const/statically initialized?

Reply:
there is `mono_mini_arch_lock ()` available in this file

Reply:
In particular, having reread the compiler developer's notes on a similar matter,
you need a barrier between reading param_reg_to_index_inited, and reading what it guards.

Like this:

```
if (!param_reg_to_index_inited) {
		for (i = 0; i < PARAM_REGS; ++i)
			general_param_reg_to_index [param_regs[i]] = i;
		for (i = 0; i < FLOAT_PARAM_REGS; ++i)
			float_param_reg_to_index [float_param_regs[i]] = i;
		mono_memory_barrier ();
		param_reg_to_index_inited = 1;
	}
else {
		mono_memory_barrier (); // << add this
}
```

Otherwise you can see uninitialized data.

We should probably use a "once" -- which indeed the Win32/arm version
has a barrier every time.

C++ initialized static locals now offer the best functionality: terse, thread safe, no barrier.
They do use a thread local under the covers, which is a bit unfortunate.

Reply:
Jepp, this was old code behaving like this so I didn't change it (added two years ago by this commit https://github.com/mono/mono/commit/a58c2ef804779be70832c6b0e092fbca1edaebf1). C++ has acquire/release semantics together with sequential consistent memory order to correctly order load/stores on different architectures and since C++11 made static locals thread safe, this would have been fine, on C++11, but now we have C and this is code running on AMD64 targets (part of the runtime and not the compiler), but in order to close this theoretical initialization race I can extend the code with the proposed memory barrier.

Reply:
C11 atomics should be easy to use and super efficient here (on non-arm at least).

Reply:
Maybe not with Visual C++ though?

Reply:
We have mono atomic's implementation that could be an option instead of barriers. Moving to static initialization could also be an option, but I think original author didn't want to manual maintain that mapping in the first place (but it should be static per ABI and should not change once set). I could give it a try as an alternative, but first I need to add support for stack parameters in DYNCALL as part of this PR (that why it's still work in progress).

Reply:
Mono atomics are probably not the same thing or as efficient, as C11 atomics (or C++ atomics<>)
I understand the desire to generate the data, the code might appear smaller than the generated data, but race conditions aren't worthwhile.
Sometimes I write a program to generate the data (printf the source) and check that in.
I wonder if we are misusing barriers elsewhere in the runtime.
Work in progress understood.

Reply:
@vargaz Any opinions related to the above discussion?

Reply:
The code in https://github.com/mono/mono/pull/9334#discussion_r198606590 looks correct to me.

Reply:
Added memory barrier as outline above.